libs/langchain-community/src/vectorstores/turbopuffer.ts

```diff
@@ -0,0 +1,213 @@
+import { type DocumentInterface, Document } from "@langchain/core/documents";
+import type { EmbeddingsInterface } from "@langchain/core/embeddings";
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import { VectorStore } from "@langchain/core/vectorstores";
+
+export interface TurbopufferHeaders {
+  headers: {
+    Authorization: string;
+    "Content-Type": string;
+  };
+}
+
+export type TurbopufferDistanceMetric = "cosine_distance" | "euclidean_squared";
+
+export type TurbopufferFilterType = Record<string, string>;
+
+export interface TurbopufferParams {
+  apiKey?: string;
+  namespace?: string;
+  distanceMetric?: TurbopufferDistanceMetric;
+  apiUrl?: string;
+}
+
+export interface TurbopufferQueryResult {
+  dist: number;
+  id: number;
+  vector: number[];
+  attributes: TurbopufferFilterType;
+}
+
+export class TurbopufferVectorStore extends VectorStore {
+  declare FilterType: TurbopufferFilterType;
+
+  get lc_secrets(): { [key: string]: string } {
+    return {
+      apiKey: "TURBOPUFFER_API_KEY",
+    };
+  }
+
+  get lc_aliases(): { [key: string]: string } {
+    return {
+      apiKey: "TURBOPUFFER_API_KEY",
+    };
+  }
+
+  // Handle minification for tracing
+  static lc_name(): string {
+    return "TurbopufferVectorStore";
+  }
+
+  protected distanceMetric: TurbopufferDistanceMetric = "cosine_distance";
+
+  protected apiKey: string;
+
+  protected namespace = "default";
+
+  protected apiUrl = "https://api.turbopuffer.com/v1/";
+
+  public _vectorstoreType(): string {
+    return "turbopuffer";
+  }
+
+  constructor(embeddings: EmbeddingsInterface, args: TurbopufferParams) {
+    super(embeddings, args);
+
+    const apiKey = args.apiKey ?? getEnvironmentVariable("TURBOPUFFER_API_KEY");
+    if (!apiKey) {
+      throw new Error(
+        [
+          "Turbopuffer API key not found.",
+          `Please pass it in as "apiKey" or set it as an environment variable called "TURBOPUFFER_API_KEY"`,
+        ].join("\n")
+      );
+    }
+    this.apiKey = apiKey;
+    this.namespace = args.namespace ?? this.namespace;
+    this.distanceMetric = args.distanceMetric ?? this.distanceMetric;
+    this.apiUrl = args.apiUrl ?? this.apiUrl;
+  }
+
+  getJsonHeader(): TurbopufferHeaders {
+    return {
+      headers: {
+        Authorization: `Bearer ${this.apiKey}`,
+        "Content-Type": "application/json",
+      },
+    };
+  }
+
+  async addVectors(
+    vectors: number[][],
```

Comment:
It's also valid to pass a `number[]`, but this may be too uncommon in JS (versus e.g. Python, where it's common)—not strictly required for my sake to include though :)